File: src/DesignPatterns/DIY/Memento/Memento.java
package DesignPatterns.DIY.Memento;

import java.util.ArrayList;
import java.util.List;

// originator
class Scenery{
    private String state;

    public Scenery(String state){
        this.state = state;
    }
    public Photograph capturePhoto(){
        return new Photograph(state);
    }
    public void restorePhoto(Photograph photo){
        state = photo.getPhoto();
    }
    public void setPhoto(String state){
        this.state = state;
    }
    public String getPhoto(){
        return state;
    }

}

// memento 
class Photograph{
    private String state;
    
    public Photograph(String state){
        this.state = state;
    }
    public String getPhoto(){
        return state;
    }
}

// caretaker
class Photographer{

    private List<Photograph> photos = new ArrayList<>();

    public void addPhoto (Photograph photo){
        photos.add(photo);
    }
    public Photograph getPhoto(int index){
        return photos.get(index);
    }
    public int getPhotosCount(){
        return photos.size();
    }

}


// main method
public class Memento {
    public static void main(String[] args) {
        // initial scenery
        Scenery scenery = new Scenery("Sunny day.");
        Photographer photographer = new Photographer();

        // setting the initial scenery as index 0
        photographer.addPhoto(scenery.capturePhoto());
        String currentPhoto = scenery.getPhoto();
        System.out.println("Current Scenery: " + currentPhoto);

        // creating a new scenery and setting it as index 1
        scenery.setPhoto("Cloudy day.");
        photographer.addPhoto(scenery.capturePhoto());
        currentPhoto = scenery.getPhoto();
        System.out.println("Current Scenery: " + currentPhoto);

        // creating a new scenery and setting it as index 1
        scenery.setPhoto("Rainy day.");
        currentPhoto = scenery.getPhoto();
        System.out.println("Current Scenery: " + currentPhoto);

        // restoring photo with the index 1.
        //output should be cloudy day as it is the value of index 1
        scenery.restorePhoto(photographer.getPhoto(1));
        currentPhoto = scenery.getPhoto();
        System.out.println("Current Scenery: " + currentPhoto);

    }
}

File: bin/DIY/Memento/Memento.ts
// Memento.ts

// Originator
class Scenery {
    private state: string;

    constructor(state: string) {
        this.state = state;
    }

    capturePhoto(): Photograph {
        return new Photograph(this.state);
    }

    restorePhoto(photo: Photograph): void {
        this.state = photo.getPhoto();
    }

    setPhoto(state: string): void {
        this.state = state;
    }

    getPhoto(): string {
        return this.state;
    }
}

// Memento
class Photograph {
    private state: string;

    constructor(state: string) {
        this.state = state;
    }

    getPhoto(): string {
        return this.state;
    }
}

// Caretaker
class Photographer {
    private photos: Photograph[] = [];

    addPhoto(photo: Photograph): void {
        this.photos.push(photo);
    }

    getPhoto(index: number): Photograph {
        return this.photos[index];
    }

    getPhotosCount(): number {
        return this.photos.length;
    }
}

// Main
(function main() {
    const scenery = new Scenery("Sunny day.");
    const photographer = new Photographer();

    photographer.addPhoto(scenery.capturePhoto());
    let currentPhoto = scenery.getPhoto();
    console.log("Current Scenery: " + currentPhoto);

    scenery.setPhoto("Cloudy day.");
    photographer.addPhoto(scenery.capturePhoto());
    currentPhoto = scenery.getPhoto();
    console.log("Current Scenery: " + currentPhoto);

    scenery.setPhoto("Rainy day.");
    currentPhoto = scenery.getPhoto();
    console.log("Current Scenery: " + currentPhoto);

    scenery.restorePhoto(photographer.getPhoto(1));
    currentPhoto = scenery.getPhoto();
    console.log("Current Scenery: " + currentPhoto);
})();